java/ql/test/library-tests/dataflow/modulus-analysis/ModulusAnalysis.expected

```diff
@@ -0,0 +1,74 @@
+| ModulusAnalysis.java:12:25:12:29 | seven | 0 | 7 | 0 |
+| ModulusAnalysis.java:12:25:12:29 | seven | SSA def(seven) | 0 | 0 |
+| ModulusAnalysis.java:13:32:13:34 | mul | 0 | 3 | 42 |
+| ModulusAnalysis.java:13:32:13:34 | mul | SSA def(mul) | 0 | 0 |
+| ModulusAnalysis.java:16:17:18:23 | ...?...:... | 0 | 3 | 4 |
+| ModulusAnalysis.java:17:15:17:15 | i | SSA init(i) | 0 | 0 |
+| ModulusAnalysis.java:17:15:17:19 | ... * ... | 0 | 0 | 4 |
+| ModulusAnalysis.java:17:15:17:23 | ... + ... | 0 | 3 | 4 |
+| ModulusAnalysis.java:17:19:17:19 | 4 | 0 | 4 | 0 |
+| ModulusAnalysis.java:17:23:17:23 | 3 | 0 | 3 | 0 |
+| ModulusAnalysis.java:18:15:18:15 | i | SSA init(i) | 0 | 0 |
+| ModulusAnalysis.java:18:15:18:19 | ... * ... | 0 | 0 | 8 |
+| ModulusAnalysis.java:18:15:18:23 | ... + ... | 0 | 7 | 8 |
+| ModulusAnalysis.java:18:19:18:19 | 8 | 0 | 8 | 0 |
+| ModulusAnalysis.java:18:23:18:23 | 7 | 0 | 7 | 0 |
+| ModulusAnalysis.java:19:28:19:28 | j | 0 | 3 | 4 |
+| ModulusAnalysis.java:19:28:19:28 | j | SSA def(j) | 0 | 0 |
+| ModulusAnalysis.java:21:13:21:13 | x | SSA init(x) | 0 | 0 |
+| ModulusAnalysis.java:21:17:21:18 | c1 | 0 | 42 | 0 |
+| ModulusAnalysis.java:21:17:21:18 | c1 | SSA init(this.c1) | 0 | 0 |
+| ModulusAnalysis.java:21:23:21:23 | 3 | 0 | 3 | 0 |
+| ModulusAnalysis.java:21:28:21:28 | y | SSA init(y) | 0 | 0 |
+| ModulusAnalysis.java:21:32:21:33 | c1 | 0 | 42 | 0 |
+| ModulusAnalysis.java:21:32:21:33 | c1 | SSA init(this.c1) | 0 | 0 |
+| ModulusAnalysis.java:21:38:21:38 | 7 | 0 | 7 | 0 |
+| ModulusAnalysis.java:22:32:22:32 | x | 0 | 3 | 42 |
+| ModulusAnalysis.java:22:32:22:32 | x | SSA init(x) | 0 | 0 |
+| ModulusAnalysis.java:22:32:22:36 | ... + ... | 0 | 10 | 42 |
+| ModulusAnalysis.java:22:32:22:36 | ... + ... | SSA init(x) | 7 | 42 |
+| ModulusAnalysis.java:22:32:22:36 | ... + ... | SSA init(y) | 3 | 42 |
+| ModulusAnalysis.java:22:36:22:36 | y | 0 | 7 | 42 |
+| ModulusAnalysis.java:22:36:22:36 | y | SSA init(y) | 0 | 0 |
+| ModulusAnalysis.java:25:13:25:13 | x | SSA init(x) | 0 | 0 |
+| ModulusAnalysis.java:25:17:25:18 | c1 | 0 | 42 | 0 |
+| ModulusAnalysis.java:25:17:25:18 | c1 | SSA init(this.c1) | 0 | 0 |
+| ModulusAnalysis.java:25:23:25:23 | 3 | 0 | 3 | 0 |
+| ModulusAnalysis.java:25:28:25:28 | y | SSA init(y) | 0 | 0 |
+| ModulusAnalysis.java:25:32:25:33 | c1 | 0 | 42 | 0 |
+| ModulusAnalysis.java:25:32:25:33 | c1 | SSA init(this.c1) | 0 | 0 |
+| ModulusAnalysis.java:25:38:25:38 | 7 | 0 | 7 | 0 |
+| ModulusAnalysis.java:26:32:26:32 | x | 0 | 3 | 42 |
+| ModulusAnalysis.java:26:32:26:32 | x | SSA init(x) | 0 | 0 |
+| ModulusAnalysis.java:26:32:26:36 | ... - ... | 0 | 38 | 42 |
+| ModulusAnalysis.java:26:32:26:36 | ... - ... | SSA init(x) | 35 | 42 |
+| ModulusAnalysis.java:26:36:26:36 | y | 0 | 7 | 42 |
+| ModulusAnalysis.java:26:36:26:36 | y | SSA init(y) | 0 | 0 |
+| ModulusAnalysis.java:29:17:29:26 | arr.length | SSA impl upd[untracked](arr.length) | 0 | 0 |
```

Comment:
@aschackmull I expected to see an `ExprBound` here, and not an `SsaBound`. How can I construct a test case for `ExprBound`?
I would need an expression that satisfies this:
```
e.(FieldRead).getField() instanceof ArrayLengthField and
not exists(SsaVariable v | e = v.getAUse())
```

Reply:
You need an expression of the form `array.length` where the qualifier `array` isn't an `SsaVariable` access, e.g. `getArray().length`.  This also holds for would-be ssa qualifiers if the expression occurs outside the CFG.

Reply:
Thanks. I'm adding a test case for this too.